tfx/examples/bert/utils/bert_tokenizer_utils.py

```diff
@@ -0,0 +1,187 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Prepressing using tensorflow_text BertTokenizer"""
+
+import tensorflow as tf
+import tensorflow_text as text
+from tensorflow.python.eager.context import eager_mode
+import tensorflow_hub as hub
+
+_CLS = '[CLS]'
+_PAD = '[PAD]'
+_SEP = '[SEP]'
+
+class SpecialBertTokenizer():
+  """ Bert Tokenizer built ontop of tensorflow_text.BertTokenizer"""
+
+  def __init__(self, model_link):
+    self._model_link = model_link
+    self._find_special_tokens()
+
+  def _find_special_tokens(self):
+    """Find the special token ID's for [CLS] [PAD] [SEP]
+
+    Since each Bert model is trained on different vocaburary, it's important
+    to find the special token index pertaining to that model.
+    Since in Transform, tensorflow_hub.KerasLayer loads a symbolic tensor, turn
+    on eager mode to get the actual vocab_file location.
+    """
+
+    with eager_mode():
+      model = hub.KerasLayer(self._model_link)
+      vocab = model.resolved_object.vocab_file.asset_path.numpy()
+      f = open(vocab, 'r')
+      self._sep_id = None
+      self._cls_id = None
+      self._pad_id = None
+      lines = f.read().split('\n')
+      for i, line in enumerate(lines):
+        if line == _PAD:
+          self._pad_id = i
+        elif line == _CLS:
+          self._cls_id = i
+        elif line == _SEP:
+          self._sep_id = i
+        if (self._pad_id is not None and
+            self._cls_id is not None and
+            self._sep_id is not None):
+          break
+
+  def tokenize_single_sentence(
+      self,
+      sequence,
+      max_len=128,
+      add_cls=True,
+      add_sep=True,
+      add_pad=True):
+    """Tokenize a single sentence according to the vocab used by the Bert model.
+
+    Add special tokens according to config.
+
+    Args:
+      sequence: Tensor of shape [batch_size, 1].
+      max_len: The number of tokens after padding and truncating.
+      add_cls: Whether to add CLS token at the front of each sequence.
+      add_sep: Whether to add SEP token at the end of each sequence.
+      add_pad: If not add_pad, would return ragged tensor instead of tensors
+        input_mask and segment_id would be none.
+
+    Returns:
+      word_id: Tokenized sequences [batch_size, max_len].
+      input_mask: Mask padded tokens [batch_size, max_len].
+      segment_id: Distinguish multiple sequences [batch_size, max_len].
+    """
+    model = hub.KerasLayer(self._model_link)
+    vocab_file_path = model.resolved_object.vocab_file.asset_path
+    tokenizer = text.BertTokenizer(vocab_file_path, token_out_type=tf.int64)
```

Comment:
you should grab the setting for 'lower_case' from the hub module as well. This could impact quality and introduce training/serving skew if there is a mismatch.

Reply:
yes!
added